### PR TITLE
Add shear initial/boundary condition specification using power law

### PR DESCRIPTION
**Pull request by gantech**
This pull request adds the capability to specify an initial/boundary condition specification using a power law. Also adds an option to limit the minimum and maximum velocity.

Attached image shows the use of the AuxFunction in a real case setting with the following inputs

`    initial_conditions:

      - user_function: ic_1
        target_name: fluid
        user_function_name:
         velocity: wind_energy_power_law
        user_function_parameters:
          velocity:
            - 2          # Coordinate direction
            - -90.0   # Offset height
            - 90.0     # Reference height
            - 0.12      # Shear exponent
            - 9.792   # Reference velocity - x component
            - 0.0       # Reference velocity - y component
            - 0.0       # Reference velocity - z component
            - 4.0       # Minimum velocity
            - 12.0     # Maximum velocity`

ShearProfile

### CODE REVIEW

File: include/user_functions/WindEnergyPowerLawAuxFunction.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+namespace nalu{
+
+
+
+/** Create power law velocity profile aux function for wind energy applications
+ *
+ *  This function is used as an initial or boundary condition,
+ *  primarily in simulation of wind turbines under specified shear
+ *  with a power law profile. The function implemented is 
+ * \f[]
```
Comment (sayerhs) [inline include/user_functions/WindEnergyPowerLawAuxFunction.h]: Typo: remove closing bracket

File: include/user_functions/WindEnergyPowerLawAuxFunction.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    const unsigned spatialDimension,
+    const unsigned numPoints,
+    double * fieldPtr,
+    const unsigned fieldSize,
+    const unsigned beginPos,
+    const unsigned endPos) const;
+  
+private:
+
+    int coord_dir_{2}; // Coordinate direction - 0/1/2
```
Comment (sayerhs) [inline include/user_functions/WindEnergyPowerLawAuxFunction.h]: The default values are never used as all values are expected to be present in the input file, are they necessary?

File: src/user_functions/WindEnergyPowerLawAuxFunction.C [hunk 3]
```diff
@@ -0,0 +1,83 @@
+/*------------------------------------------------------------------------*/
+/*  Copyright 2019 National Renewable Energy Laboratory.                  */
+/*  This software is released under the license detailed                  */
+/*  in the file, LICENSE, which is located in the top-level Nalu          */
+/*  directory structure                                                   */
+/*------------------------------------------------------------------------*/
+
+#include "user_functions/WindEnergyPowerLawAuxFunction.h"
+#include <algorithm>
```
Comment (sayerhs) [inline src/user_functions/WindEnergyPowerLawAuxFunction.C]: Unnecessary header?

File: src/user_functions/WindEnergyPowerLawAuxFunction.C [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+// basic c++
+#include <cmath>
+#include <vector>
+#include <stdexcept>
```
Comment (sayerhs) [inline src/user_functions/WindEnergyPowerLawAuxFunction.C]: unnecessary header?

File: src/user_functions/WindEnergyPowerLawAuxFunction.C [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+WindEnergyPowerLawAuxFunction::do_evaluate(
+  const double *coords,
+  const double /*time*/,
+  const unsigned /*spatialDimension*/,
+  const unsigned numPoints,
+  double * fieldPtr,
+  const unsigned fieldSize,
+  const unsigned /*beginPos*/,
+  const unsigned /*endPos*/) const
+{
```
Comment (sayerhs) [inline src/user_functions/WindEnergyPowerLawAuxFunction.C]: The floating point ops in this function could be optimized? It appears to be doing the same calculations again and again. `uref_` vector could be normalized in constructor and used here?

File: src/LowMachEquationSystem.C [hunk 6]
```diff
@@ -620,6 +622,16 @@ LowMachEquationSystem::register_initial_condition_fcn(
...
       }
     }
+    else if ( fcnName == "wind_energy_power_law") {
+        auto it = theParams.find(dofName);
+        if (it != theParams.end()) {
+            auto& fp = it->second;
+            theAuxFunc = new WindEnergyPowerLawAuxFunction(0,nDim,fp);
+        }
+        else {
+            throw std::runtime_error("Widn Energy Power Law aux function missing parameters in initial condition");
```
Comment (jrood-nrel) [inline src/LowMachEquationSystem.C]: Typo

**Comment by gantech** [comment]
Closing pull request for now. Will address these comments and make a new one.

**Pull request closed without merging**